### ISSUE TITLE
Detach method missingin IntegratorTwoStep

### ISSUE THREAD

**Issue by YouriRan** [glotzerlab/hoomd-blue#1125]
## Description

Detach method missing in IntegratorTwoStep leading to error when using intergrator.methods.remove or setting to new list. 

## Script

```python
import hoomd
import numpy as np
import gsd.hoomd
import math
import itertools

# set up box
m = 8
N_particles = m**3
spacing = 2.0
K = math.ceil(N_particles**(1/3))
L = K*spacing
x = np.linspace(-L/2+1, L/2-1, K, endpoint=False)
position = list(itertools.product(x, repeat=3))

snapshot = gsd.hoomd.Snapshot()
snapshot.particles.N = N_particles
snapshot.particles.position = position[0:N_particles]
snapshot.particles.typeid = [0] * N_particles
snapshot.configuration.box = [L, L, L, 0, 0, 0]
snapshot.particles.types = ['A']

with gsd.hoomd.open(name='init.gsd', mode='wb') as f:
    f.append(snapshot)

# Initialize potential
cell = hoomd.md.nlist.Cell()
lj = hoomd.md.pair.LJ(nlist=cell)

lj.params[('A','A')] = {'epsilon':1.0, 'sigma':1.0}
lj.r_cut[('A','A')] = 3.0

# set up simulation
cpu = hoomd.device.CPU()
sim = hoomd.Simulation(device=cpu, seed=12)
sim.create_state_from_gsd(filename='init.gsd')
sim.state.thermalize_particle_momenta(filter=hoomd.filter.All(), kT=1.2)

trj_trigger = hoomd.trigger.Periodic(1000)
trj = hoomd.write.GSD(filename='out.gsd', trigger=trj_trigger, mode='wb')
sim.operations.writers.append(trj)

dt = 1e-3
nvt1 = hoomd.md.methods.NVT(filter=hoomd.filter.All(), kT=1.2, tau=0.1)
nvt2 = hoomd.md.methods.NVT(filter=hoomd.filter.All(), kT=1.5, tau=0.1)

integrator = hoomd.md.Integrator(dt=dt, forces=[lj], methods=[nvt1])
sim.operations.integrator = integrator
sim.run(1e4)

def option1():
    integrator.methods = [nvt2]
    sim.run(1e4)

def option2():
    integrator.methods.remove(nvt1)
    integrator.methods.append(nvt2)
    sim.run(1e4)

option1()

```

## Output

```
Traceback (most recent call last):
  File "/home/youriran/pls/sims/test/test.py", line 71, in <module>
    option1()
  File "/home/youriran/pls/sims/test/test.py", line 63, in option1
    integrator.methods = [nvt2]
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/operation.py", line 68, in __setattr__
    self._setattr_hook(attr, value)
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/operation.py", line 72, in _setattr_hook
    super().__setattr__(attr, value)
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/md/integrate.py", line 92, in methods
    _set_synced_list(self._methods, value)
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/md/integrate.py", line 29, in _set_synced_list
    old_list.clear()
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/_collections_abc.py", line 1078, in clear
    self.pop()
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/_collections_abc.py", line 1100, in pop
    del self[index]
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/data/syncedlist.py", line 148, in __delitem__
    self._detach_value(old_value)
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/data/syncedlist.py", line 213, in _detach_value
    value._detach()
  File "/home/youriran/miniconda3/envs/hoomdbeta008/lib/python3.9/site-packages/hoomd/operation.py", line 233, in _detach
    self._cpp_obj.notifyDetach()
AttributeError: 'hoomd.md._md.TwoStepNVTMTK' object has no attribute 'notifyDetach'
```

## Expected output

Method removed and replaced by new method, sim running with two steps.

## Configuration

OS: Ubuntu 21.04
CPU: Intel i5
HOOMD version: 3.0.0dev8
Build: conda cpu_py39h85fc912_0
Python version: 3.9.6

## Developer

I found that adding a notifyDetach() function, similar to that found in Compute.cc and Compute.h, to IntegratorTwoStep.cc and IntegratorTwoStep.h resolves the issue. I can make a pull request.

**Comment by joaander** [comment]
Thanks for the report. I would think that the correct solution would be to add `notifyDetach` to `IntegrationMethodTwoStep`. `IntegratorTwoStep` should inherit it from `Updater`. And yes, a pull request to fix this is welcome.

I don't understand why the test on #1118 or this test donot trigger this error:
https://github.com/glotzerlab/hoomd-blue/blob/0d10bef3376cdeb6538005e74d1207ff8107d5bd/hoomd/md/pytest/test_methods.py#L180-L199
@b-butler any ideas?

In  your pull request, please modify or add a new test to `test_methods.py` that fails before your change and passes after.

Also be aware that we just recently merged a fix to a bug that caused `remove` to remove other methods than intended: #1114. This fix is not yet in a tagged release.

**Comment by YouriRan** [comment]
This issue is resolved by [#1114](https://github.com/glotzerlab/hoomd-blue/pull/1114); compiling from source doesn't give me this issue. Thanks for that! Also I see in the output of pytest that test_methods.py::test_switch_methods is being called and with success.

**Comment by joaander** [comment]
OK, now I see. #1053 solves this by defining a custom `Method._detach` method that doesn't call the base class `_detach` - thus `notifyDetach` is not called for integration methods.
